draft-thomson-train-protocol.md

```diff
@@ -336,6 +336,26 @@ same datagram SHOULD ignore any rate limit signal.  Such a datagram might be
 entirely spoofed.
 
 
+## Feedback To Sender About Signals {#feedback}
+
+Information about rate limits is intended for the sending application.  Any
+signal from network elements can be propagated to the receiving application
+using an implementation-defined mechanism.
+
+This document does not define a means for indicating what was received.
+Feedback can be implemented in a manner that best suits applications
+requirements.
+
+A sender can use any acknowledgment mechanism provided by the QUIC version in
+use to learn whether datagrams containing TRAIN packets were received.  This
+might inform whether to send additional TRAIN packets in the event that a
+datagram is lost.
+
```

Comment:
I am not a great fan of that solution. It assumes that senders keep track of which packet was coalesced with what other packet, and that receiving one implies receiving the other. But that's not a guarantee. For example, it could happen that the TRAIN packet is corrupted, but the next packet in the datagram is received just fine.

Reply:
A corrupted TRAIN packet is unlikely to be easy to discard.  Assuming that it is, then that is fine.

To answer your question about end-to-end, my argument is very specifically about end-to-end.  If the message does not hit the application, then it is middle-to-middle.  ECN differs because it is a message for the congestion controller.  This message is for the rate adaptation mechanisms in the application, not the QUIC stack.

I can see why you would want to close the loop at the QUIC layer.  It's certainly possible.  I'm not opposed to it, I just prefer not to start out by specifying it.

The main case that I think might inform this decision is are HLS, Dash, and analogous video streaming protocols.  In those cases, the rate adaptation occurs by requesting a different quality level (or a rung on a ladder).  This is done without modifying the application at all, because rate adaptation is receiver driven in those protocols.  Having a potentially conflicting signal at the QUIC layer isn't helpful in those cases.  Given that these are the primary driver for this design, that weakens the case for a QUIC signal.

Reply:
I see the case for "application signal". Perhaps we need to make it explicitly in the draft?

Reply:
I agree that the "ack of coalesced packet" works in practice. I can swallow hard and live with that. But then, do we even need to mention it? Defining it sort of implies that the transport will issue a new TRAIN if the first one was not acked. But if the signal is truly end-to-end, the decision to send the TRAIN packet probably belongs to the application.

Reply:
Ah, good point.  Maybe we should instead say:

> This might help inform whether to send additional TRAIN packets in the event that a datagram is lost. However, rather than relying on transport signals, an application might be better able to indicate what has been received and processed.

Reply:
Yes, that would do it. Also, "TRAIN packets could be stripped from datagrams in the network"... which means the datagram would be acknowledged, and nobody would be any wiser.